### PR TITLE
🎉 Source Salesforce: Add the ability to filter streams

### LINKED ISSUE #7031: Salesforce: investigate selecting schemas/streams before the discovery step
## Tell us about the problem you're trying to solve
Users of the Salesforce connector who have a lot of tables in their salesforce instance (1k+) find the connector difficult to use because it returns all the tables, causing the following issues: 

1. discover operation is too slow when there are many tables
2. the UI is very difficult to use because there

in many cases it is unnecessary to pull all the data because the user only needs to select a few tables. 

## Describe the solution you’d like
as a user I would like some way of filtering the data during the config stage. Is it possible to filter by a "schema" (like in Postgres) or maybe insert a regex? I don't know what options are available from the API but would like to solve this problem somehow. 

## Acceptance Criteria
The output of this ticket should be a proposal of the options available to us to solve the problem above. After the proposal is reviewed by Airbyte, create tickets to implement the solution.

### PR DESCRIPTION
**Pull request by yevhenii-ldv**
## What
*resolves #7031.*

## How
*A function has been added to the specification that allows users to specify values for filtering streams, as well as specify the type of option for filtering.
It is **Optional** field*

## Recommended reading order
1. `x.java`
2. `y.python`

#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

### CODE REVIEW

File: airbyte-integrations/connectors/source-salesforce/unit_tests/unit_test.py [hunk 1]
```diff
@@ -298,3 +298,30 @@ def test_download_data_filter_null_bytes(stream_bulk_config, stream_bulk_api):
...
+)
+def test_discover_with_streams_criteria_param(streams_criteria, number_of_filtered_streams, stream_rest_config):
+    updated_config = {**stream_rest_config, **{"streams_criteria": streams_criteria}}
+    sf_object = Salesforce(**stream_rest_config)
+    sf_object.login = Mock()
+    sf_object.access_token = Mock()
+    sf_object.instance_url = "https://fase-account.salesforce.com"
+    sf_object.describe = Mock(return_value={"sobjects": [{"name": f"FakeStream{i}"} for i in range(100)]})
+    filtered_streams = sf_object.get_validated_streams(config=updated_config)
+    assert len(filtered_streams) == number_of_filtered_streams
```
Comment (htrueman) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/unit_test.py:327]: perhaps we may check the steams by name instead of just count them?

**Review by htrueman — COMMENTED** [review]
LGTM. Have just one question for unit tests.

File: airbyte-integrations/connectors/source-salesforce/unit_tests/unit_test.py (hunk 1 above)
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/unit_test.py:327]: That is, you propose to use the same method to filter the names of streams, and in the end compare them with each other, it is strange?
Could you tell us in more detail how you see this implementation?

Comment (htrueman) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/unit_test.py:327]: I just suggest to compare `filtered_stream == expected_filtered_streams`.
E.g. `['stream1', 'stream2'] == ['stream1']`
Is it possible?

**Review by htrueman — APPROVED** [review]

**Review by Zirochkaa — APPROVED** [review]

**Pull request merged**